Demonstrate using TAXRANK for ranks and properties

cthoyt opened the pull request:
Closes #96, closes #97

This PR does the following (updated Feb 19, 2025):

1. Adds triples connecting NCBITaxon terms to TAXRANK terms using the TAXRANK:100000 (has rank) relation. This duplicates, in parallel, the existing ad-hoc `ncbitaxon:has_rank` relations that point to ad-hoc rank terms in the `NCBITaxon:` namespace
2. Explicitly handles the `no rank` rank
3. Adds deprecation notice to the predicate `ncbitaxon:has_rank`, removes logical axioms, updates its name, and adds a replaced-by notice pointing to TAXRANK:100000
4. Adds a deprecation notice to all ad-hoc NCBITaxon ranks with appropriate replaced-by notices pointing to TAXRANK terms

## Related Work

This PR depends on https://github.com/phenoscape/taxrank/pull/5, where I added the remaining 12 ranks that weren't already represented. That PR was merged and released on November 28<sup>th</sup>, 2024. Related: that PR included adding a tax rank for "strain", which was the point of discussion also in #107.

This PR also depends on https://github.com/phenoscape/taxrank/pull/10 and https://github.com/phenoscape/taxrank/pull/11, where I added 5 more ranks that weren't represented. Those PRs were merged on September 24<sup>th</sup>, 2025.

matentzn reviewed (changes requested):
I personally very much like this change, and I approve of it, but I will request changes so that it is not accidentally merged before I can get some community feedback (and we can use my disapproval as a blocker until the taxrank PR is released).

Thank you very much!

jplfaria commented:
@matentzn, are there any updates on "community feedback" for this change? Is this issue the best place to provide feedback? 

I have been working to create ontology descriptions of other taxonomy sources (gtdb and silva) where I include TAXRANK per @cthoyt suggestion. Using TAXRANK to define taxonomic ranks will vastly facilitate comparison and mapping between different taxonomies.

cmungall commented:
I’m I favor of this change

On Sun, Feb 16, 2025 at 7:56 PM jplfaria ***@***.***> wrote:

> @matentzn <https://github.com/matentzn>, are there any updates on
> "community feedback" for this change? Is this issue the best place to
> provide feedback?
>
> I have been working to create ontology descriptions of other taxonomy
> sources (gtdb and silva) where I include TAXRANK per @cthoyt
> <https://github.com/cthoyt> suggestion. Using TAXRANK to define taxonomic
> ranks will vastly facilitate comparison and mapping between different
> taxonomies.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/obophenotype/ncbitaxon/pull/120#issuecomment-2661910664>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMMOPXHBWA4EYBCPP55G32QFMWDAVCNFSM6AAAAABSTUPRXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRHEYTANRWGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
> [image: jplfaria]*jplfaria* left a comment (obophenotype/ncbitaxon#120)
> <https://github.com/obophenotype/ncbitaxon/pull/120#issuecomment-2661910664>
>
> @matentzn <https://github.com/matentzn>, are there any updates on
> "community feedback" for this change? Is this issue the best place to
> provide feedback?
>
> I have been working to create ontology descriptions of other taxonomy
> sources (gtdb and silva) where I include TAXRANK per @cthoyt
> <https://github.com/cthoyt> suggestion. Using TAXRANK to define taxonomic
> ranks will vastly facilitate comparison and mapping between different
> taxonomies.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/obophenotype/ncbitaxon/pull/120#issuecomment-2661910664>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMMOPXHBWA4EYBCPP55G32QFMWDAVCNFSM6AAAAABSTUPRXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRHEYTANRWGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

jamesaoverton commented:
We discussed this in the OBO Community Slack on November 27, but unfortunately the discussion did not get back here.

The current version of this PR drops the `ncbitaxon:has_rank` predicate that we've been using for 20+ years, and so breaks any system that queries for or displays the old ranks.

If the PR just added new TAXRANK stuff without dropping the old stuff, it wouldn't be a breaking change.

cmungall commented:
I think having an interim period with both would be fine, but can we start a survey of people who are using the old properties with a view to declaring EOL?

We use ncbitaxon heavily but this change would not break things for us AFAICT, but others use cases may differ

cthoyt commented:
It's not exactly clear what are the accept conditions for this PR, but I made the following updates:

1. Reverted deletions of old code that created old-style ad-hoc identifiers
2. Added deprecations and replaced-by relations to old-style ad-hoc identifiers
3. Add all usages of the TAXRANK relation and TAXRANK terms in parallel.

Therefore, you only have to judge if you think it's alright to duplicate what's here already (in two different flavors)

Otherwise, please give explicit instructions on what to change to get this accepted

jamesaoverton commented:
1a48d43 addresses my concerns about breakage.

In 27ecf11 I added "obsolete" to labels and removed subClassOf assertions, following OBO obsoletion best practices, drafted as principle 19 here: https://github.com/OBOFoundry/OBOFoundry.github.io/blob/nataled-patch-50/principles/fp-019-term-stability.md#implementation

We could also add an IAO:0000231 'obsolescence reason', but the IAO:0100001 'term replaced by' is doing the real work.

We don't often deprecate annotation properties, and that draft principle doesn't mention this case. It's reasonable to drop the old annotations after some deprecation period an giving ample notice. https://obofoundry.org/principles/fp-013-notification.html has some discussion, but the requirements there are minimal. NCBI Taxonomy is one of the oldest and most-used OBO projects, so I think extra care is justified.

matentzn reviewed (approved):
Approving based on @cmungall and @jamesaoverton approval; did _not_ do a test or deep dive of the code (I assume you did that @anitacaron :D). Thanks all for the patience.

cthoyt committed:
Use TAXRANK for ranks and properties

Closes #96, closes #97

cthoyt commented:
@matentzn it looks like chris and james both agree on this, and I have tested it locally with the help of Jim. We have community consensus, is there any other blocker for merging this?

matentzn commented:
> @matentzn I think both Chris (comment here https://github.com/obophenotype/ncbitaxon/pull/120#issuecomment-2666312168) and James (comment here https://github.com/obophenotype/ncbitaxon/pull/120#issuecomment-2674834029) agreed that having them in parallel is a good thing that will support the community to transition to the new ones.

Sorry I was grammatically off. What I meant to say (thats why I approved ;) THAT item was the only thing I cared about significantly here! So, I am in full agreement of this PR.

anitacaron commented:
I've checked the code, and ideally, the axioms for TAXRANK should be sourced from http://purl.obolibrary.org/obo/taxrank.owl.

It seems there's still missing some ranks:

```
Summary of unrecognized ranks:

Counter({'realm': 7, 'domain': 3, 'acellular root': 1, 'cellular root': 1, 'subvariety': 1})
```

cthoyt commented:
@anitacaron it appears this has popped up since NCBITaxon has updated in the very long time of review. This requires coordinating with TAXRANK ontology, and is not related for this PR. I promise I will take care of it, but I would rather not have a moving target on this once since it has been very difficult to get additional rounds of feedback.

matentzn reviewed (changes requested):
Lets do this the right way - get the ranks into taxrank, then merge this PR; 

@cthoyt let me know if you need help getting the ranks in!

cthoyt committed:
Add missing ranks following improvements to TAXRANK

https://github.com/phenoscape/taxrank/pull/10 and https://github.com/phenoscape/taxrank/pull/11

cthoyt committed:
Reduce diff based on no rank

cthoyt committed:
Cleanup script

cthoyt committed:
Add switch on parents in TAXRANK

cthoyt committed:
Merge remote-tracking branch 'upstream/master' into pr/120

cthoyt committed:
Update ncbitaxon.py

cthoyt commented:
#136 was just merged and I've taken care of rebasing this PR on top of that (no changes were necessary since this PR also was already incorporating the new taxranks)

matentzn reviewed (approved):
This is great. THANKS @cthoyt for going the long and arduous path to do this right! Thanks to all the reviewers.

I will leave this merge to @anitacaron as the steward of this ontology.

cthoyt commented:
thanks @anitacaron
